### PR TITLE
MAMA/MAMACPP/MAMAJAVA/MAMADOTNET: Publisher events

### PR DESCRIPTION
**Pull request by reed-jpm-alpert**
Full set of changes across C/C++/Java/C#.

Changes to Java unit test to support bridge/tport on cmd line.
Changes to Scons in windows to allow dotnet unit tests to build.
Tabs converted to 4 spaces, but this also found pre-existing tabs.
Some changes to allow C code to compile on Visual Studio - requires all vars declared at top of block.
Includes changes to bridge interface.
Tested with reflection bridge that uses mamaFieldCache as a LVC.

Signed-off-by: Reed Alpert <reed.alpert@gmail.com>

[Review on Reviewable](https://reviewable.io/reviews/openmama/openmama/100)

### CODE REVIEW

**Comment by reed-jpm-alpert** [comment]
Added examples for C/C++/Java/C#.

File: mama/c_cpp/src/c/clientmanageresponder.c [hunk 1]
```diff
@@ -90,231 +90,234 @@ static void commandDestroyCB (wList list, void *element, void *closure)
...
-		/* Free the impl itself. */
-		free(impl);
-	}
+        /* Free the impl itself. */
+        free(impl);
+    }
 
-	return ret;
+    return ret;
 }
```
Comment (fquinner) [inline mama/c_cpp/src/c/clientmanageresponder.c]: Can you revert the whitespace changes in the function above here please? I'm OK with fixing indentation issues in functions which are being modified at the time but there doesn't seem to be any functional changes in the code above so fixing indentation in this instance just makes this hard to review. I'll make an exception if you can point to a change that has been made somewhere in here functionally that I have missed.

File: mama/c_cpp/src/c/clientmanageresponder.c [hunk 2]
```diff
@@ -90,231 +90,234 @@ static void commandDestroyCB (wList list, void *element, void *closure)
...
+            int nextTransportIndex                = 0;
+            mamaPublisher *nextPublisher        = impl->publishers;
+            mamaSubscription *nextSubscription    = impl->subscriptions;    
+            while((ret == MAMA_STATUS_OK) && (nextTransportIndex < impl->numberTransportBridges))
+            {
+                /* The publisher allows commands to respond to point to point requests
+                 * without creating a new publisher, this must be created using the correct
+                 * transport bridge.
+                 */
+                ret = mamaPublisher_createByIndex(
```
Comment (fquinner) [inline mama/c_cpp/src/c/clientmanageresponder.c]: I was about to complain that this function is public and therefore you shouldn't change it but it appears that is it not, despite it's public-like name. While you're changing the signature to that function which will break any other implementation that depend on this function anyway, could you refactor it to be named mamaPublisherImpl_createByIndex as well please just to make it a little obvious that this is an internal function?

File: mama/c_cpp/src/c/clientmanageresponder.c [hunk 3]
```diff
@@ -90,231 +90,234 @@ static void commandDestroyCB (wList list, void *element, void *closure)
...
 
-/*	Description	:	This function will create the cm responder, the responder will listen for management messages from
- *					the currently active feed handlers. This is done by creating a publisher and a subscription for each
- *					sub transport bridge in the main transport.
- *	Arguments	:	responder				[O] To return the new responder.
- *					tport					[I] The main transport.
- *					numberTransportBridges	[I] The number of transport bridges.
- *	Returns		:	MAMA_STATUS_NOMEM
- *					MAMA_STATUS_OK
+/*    Description    :    This function will create the cm responder, the responder will listen for management messages from
```
Comment (fquinner) [inline mama/c_cpp/src/c/clientmanageresponder.c]: Again, can you revert all whitespace corrections between here and the end of the file please?

File: common/c_cpp/src/c/mempool.c [hunk 4]
```diff
@@ -76,6 +76,7 @@ memoryNode*
 memoryPool_addNode (memoryPool* pool, size_t nodeSize)
```
Comment (fquinner) [inline common/c_cpp/src/c/mempool.c]: Mempool mixed declaration changes are not relevant to this change - please remove from pull request and if necessary raise a separate pull request for that.

File: mama/c_cpp/src/c/listenermsgcallback.c [hunk 5]
```diff
@@ -86,11 +86,13 @@ listenerMsgCallback_create( listenerMsgCallback *result,
     msgCallback* callback = (msgCallback*)calloc( 1, sizeof( msgCallback ) );
 
 #ifdef WITH_ENTITLEMENTS  /* No listener creation without a client. */
```
Comment (fquinner) [inline mama/c_cpp/src/c/listenermsgcallback.c]: Please revert all changes in this file or we'll get merge conflicts when the entitlements stuff lands, and it's not relevant to publisher events.

File: mama/c_cpp/src/c/mama.c [hunk 6]
```diff
@@ -1640,7 +1640,7 @@ mdrvImpl_ParseServersProperty()
 
     if (gServerProperty == NULL)
     {
-        memset (gServers, 0, sizeof(gServers));
+        memset ((char*) gServers, 0, sizeof(gServers));
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama.c]: Again please revert - not relevant to publisher events changes.

File: mama/c_cpp/src/c/mama/middleware.h [hunk 7]
```diff
@@ -51,7 +51,8 @@ typedef enum mamaMiddleware_
     MAMA_MIDDLEWARE_LBMKOMODO = 15,
     MAMA_MIDDLEWARE_UMDSKOMODO = 16,
     MAMA_MIDDLEWARE_HMS     = 17,
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama/middleware.h]: Again please revert - not relevant to publisher events changes.

File: mama/c_cpp/src/c/mama/publisher.h [hunk 8]
```diff
@@ -154,8 +232,8 @@ mamaPublisher_sendFromInbox (mamaPublisher publisher,
 MAMAExpDLL
 extern mama_status
 mamaPublisher_sendReplyToInboxHandle (mamaPublisher publisher,
-                                	  mamaMsgReply replyAddress,
-                                	  mamaMsg reply);
+                                      mamaMsgReply replyAddress,
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: Fix type / parameter alignment here please

File: mama/c_cpp/src/c/mama/publisher.h [hunk 9]
```diff
@@ -196,8 +283,61 @@ extern mama_status
...
+
+/**
+ * Retrieve the publisher callbacks.
+ *
+ * @param publisher The mamaPublisher from which to send the response.
+ * @param cb Pointer to a cb structure to be populated with the current callbacks.
+ */
+MAMAExpDLL
+extern mama_status
+mamaPublisher_getUserCallbacks (mamaPublisher publisher,
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: Fix type / parameter alignment here please

File: mama/c_cpp/src/c/middleware.c [hunk 10]
```diff
@@ -83,6 +83,9 @@ mamaMiddleware_convertFromString (const char*  str)
     if (strcasecmp (str, "hms") == 0)
         return MAMA_MIDDLEWARE_HMS;
     
+    if (strcasecmp (str, "reflect") == 0)
```
Comment (fquinner) [inline mama/c_cpp/src/c/middleware.c]: Please revert all changes in this file for the pull request branch - not relevant to change.

File: mama/c_cpp/src/c/publisher.c [hunk 11]
```diff
@@ -118,17 +176,54 @@ _createByIndex (mamaPublisher*              result,
...
+    cb->onError = publisherCallbacks ? impl->mUserCallbacks.onError : NULL;
+    cb->onDestroy = mamaPublisher_onPublisherDestroyed;        /* intercept onDestroy always to track state */
+    if (MAMA_STATUS_OK!=(status=(bridgeImpl->bridgeMamaPublisherSetUserCallbacks (
+                                impl->mMamaPublisherBridgeImpl,
+                                impl->mQueue,
+                                cb,
+                                impl->mClosure
+))))
     {
+        if (NULL != impl->mRoot) free((void*) impl->mRoot);
```
Comment (fquinner) [inline mama/c_cpp/src/c/publisher.c]: Maybe use mamaPublisherImpl_cleanup ?

File: mama/c_cpp/src/c/publisher.c [hunk 12]
```diff
@@ -611,3 +719,280 @@ mamaPublisher_getTransport (mamaPublisher   publisher,
...
+}
+
+static void mamaPublisherImpl_setState(mamaPublisherImpl* impl, mamaPublisherState state)
+{
+    /* Set the state using an atomic operation so it will be thread safe. */
+    wInterlocked_set(state, (wInterlockedInt*) &impl->mState);
+    if (mama_getLogLevel() >= MAMA_LOG_LEVEL_FINEST)
+    {
+        const char* source = "(none)";
+        const char* symbol = "(none)";;
```
Comment (fquinner) [inline mama/c_cpp/src/c/publisher.c]: Remove double semi colon here

File: mama/c_cpp/src/c/publisher.c [hunk 13]
```diff
@@ -611,3 +719,280 @@ mamaPublisher_getTransport (mamaPublisher   publisher,
...
+
+    if (impl->mBridgeImpl &&
+            impl->mMamaPublisherBridgeImpl)
+    {
+        if (MAMA_STATUS_OK!=(status=
+                    impl->mBridgeImpl->bridgeMamaPublisherDestroy
+                    (impl->mMamaPublisherBridgeImpl)))
+        {
+           mama_log (MAMA_LOG_LEVEL_ERROR, "mamaPublisher_destroy(): "
+                     "Could not destroy publisher bridge.");
```
Comment (fquinner) [inline mama/c_cpp/src/c/publisher.c]: If you're going to assign the status here, you're best using it in the log message too with a stringForStatus call.

File: mama/c_cpp/src/c/subscription.c [hunk 14]
```diff
@@ -452,6 +452,7 @@ mamaSubscription_setupBasic (
     }
 
 #ifdef WITH_ENTITLEMENTS
+	{
```
Comment (fquinner) [inline mama/c_cpp/src/c/subscription.c]: Again, please remove all these WITH_ENTITLEMENTS parenthesis changes for this PR

File: mama/c_cpp/src/c/subscription.c [hunk 15]
```diff
@@ -1194,9 +1202,11 @@ mamaSubscription_getSubjectContext (mamaSubscription subscription,
         msgUtils_getIssueSymbol (msg, &issueSymbol);
         context->mSymbol = copyString (issueSymbol);
         #ifdef WITH_ENTITLEMENTS
+		{
```
Comment (fquinner) [inline mama/c_cpp/src/c/subscription.c]: More WITH_ENTITLEMENTS  stuff - please remove along with unnecessary whitespace change at the end of this function.

File: mama/c_cpp/src/c/subscription.c [hunk 16]
```diff
@@ -1550,31 +1560,31 @@ mama_status mamaSubscription_deactivate_internal(mamaSubscriptionImpl *impl)
 
 void MAMACALLTYPE mamaSubscription_DestroyThroughQueueCB(mamaQueue Queue, void* closure)
```
Comment (fquinner) [inline mama/c_cpp/src/c/subscription.c]: Please revert all changes in this file from here on down (again, unless I have missed some functional change somewhere which is relevant to publisher events.

File: mama/c_cpp/src/c/subscriptionimpl.h [hunk 17]
```diff
@@ -221,7 +221,9 @@ mamaSubscription_getSubscriptionBridge(
  *              MAMA_STATUS_NULL_ARG
  *              MAAM_STATUS_OK
  */              
-mama_status mamaSubscription_getExpectingInitial(
+MAMAExpDLL
+extern mama_status
+mamaSubscription_getExpectingInitial(
```
Comment (fquinner) [inline mama/c_cpp/src/c/subscriptionimpl.h]: Is this required?

File: mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Wombat 
+{
+	class MAMACPPExpDLL MamaPublisherCallback
+	{
+	public:
+		MamaPublisherCallback () {}
+		virtual ~MamaPublisherCallback () {}
+		
+		virtual void onCreate (
```
Comment (fquinner) [inline mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h]: Can you fix indentation in this class for consistency?

File: mama/c_cpp/src/c/mama/publisher.h [hunk 19]
```diff
@@ -196,8 +283,61 @@ extern mama_status
 mamaPublisher_getTransport (mamaPublisher   publisher,
                             mamaTransport*  transport);
 
+/**
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: Is there any need to have _allocate and _deallocate functions for this? The struct is in a public header so users of mamaPublisherCallbacks can easily allocate it themselves or indeed (more likely) just use a stack allocated struct instead and prevent heap allocation and cleanup altogether (looks like function pointers get copied in the implementations where it's used anyway)

File: mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp [hunk 20]
```diff
@@ -60,16 +84,36 @@ static void parseCommandLine (int argc, char** argv)
...
+        {
+            gMiddleware = argv[i+1];
+            i += 2;
+        }
+        else if (strcmp ("-p", argv[i]) == 0)
+        {
+            gPayload = argv[i+1];
+            i += 2;
+        }
+        else if (strcmp ("-tport", argv[i]) == 0)
```
Comment (fquinner) [inline mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp]: Do these parameters need to be run time configurable? The approach for unit tests to date has typically been for hard-coded values within the unit test itself to be used so that users don't have to memorize the correct command line to run the unit tests. The only parameters which (should) exist at the moment  are ones which control or are dependent on 'what' is tested as opposed to 'how' it's tested. I expect the global defaults you have already specified will be fine on their own.

File: mama/c_cpp/src/gunittest/c/publishertest.cpp [hunk 21]
```diff
@@ -105,15 +159,18 @@ TEST_F (MamaPublisherTestC, CreateDestroy)
  *
  *  Expected Result: MAMA_STATUS_OK
  */
-TEST_F (MamaPublisherTestC, DISABLED_Send)
+TEST_F (MamaPublisherTestC, Send)
```
Comment (fquinner) [inline mama/c_cpp/src/gunittest/c/publishertest.cpp]: Can you add 'Event' prefix to all unit tests which depend on publisher events to be implement in the bridge (E.g. 'EventSend' here) so that it can be easily filtered out for bridges which don't support this functionality?

File: mama/dotnet/src/cs/MamaPublisher.cs [hunk 22]
```diff
@@ -128,7 +129,66 @@ protected override MamaStatus.mamaStatus DestroyNativePeer()
...
+            MamaTransport transport,
+            MamaQueue queue,
+            MamaPublisherCallback callback,
+            Object closure,
+            string topic,
+            string source,
+            string root)
+        {
+#if MAMA_WRAPPERS_CHECK_ARGUMENTS
+			if (transport == null)
```
Comment (fquinner) [inline mama/dotnet/src/cs/MamaPublisher.cs]: Can you remove tab indentations here?

File: mama/jni/src/com/wombat/mama/MamaSubscription.java [hunk 23]
```diff
@@ -81,23 +81,23 @@ public MamaSubscription()
...
-            }
-        }
-
-        finally
-        {
-            // Call the base class finalizer
-            super.finalize();
-        }
-    }
+    // protected void finalize() throws Throwable
```
Comment (fquinner) [inline mama/jni/src/com/wombat/mama/MamaSubscription.java]: Doesn't seem relevant to this change - can you remove this please?

**Comment by fquinner** [comment]
Hi Reed,

Thanks for this, finally got a chance to go through this in detail. Several changes commented throughout listed above - mostly minor and aesthetic. It also doesn't look like it will currently merge with our next branch, so I would appreciate it if you could update your fork's feature branch with the latest changes from next as well as implement the suggested changes, then we should be ready to drop it into next.

Cheers,
Frank

File: mama/c_cpp/src/c/clientmanageresponder.c (hunk 1 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/clientmanageresponder.c]: There were lots of tabs in this file, which got changed to 4 spaces when I did global search/replace. I restored the tab on this line which should make the indent revert to original.

File: mama/c_cpp/src/c/clientmanageresponder.c (hunk 3 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/clientmanageresponder.c]: I put the tabs back for the lines designated above.

File: common/c_cpp/src/c/mempool.c (hunk 4 above)
Comment (reed-jpm-alpert) [inline common/c_cpp/src/c/mempool.c]: Without this change I can't compile/test on Windows (prev pull was not Windows compat). Do you want me to break this into 2 branches, one for Windows compat and the other for publisher events?

File: mama/c_cpp/src/c/listenermsgcallback.c (hunk 5 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/listenermsgcallback.c]: Done

File: mama/c_cpp/src/c/mama.c (hunk 6 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama.c]: Done

File: mama/c_cpp/src/c/mama/middleware.h (hunk 7 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/middleware.h]: Done

File: mama/c_cpp/src/c/mama/publisher.h (hunk 8 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/publisher.h]: This change fixes the indentation to be consistent with the rest of the file, do you want me to put the tabs back so the indentation is as it was before (in red above) ?

File: mama/c_cpp/src/c/mama/publisher.h (hunk 9 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/publisher.h]: For getUserCallbacks() the alignment is consistent with the rest of the file, and this is a new method so I can't revert to a previous alignment. Not sure what you are asking for here.

File: mama/c_cpp/src/c/middleware.c (hunk 10 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/middleware.c]: Done

File: mama/c_cpp/src/c/publisher.c (hunk 11 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/publisher.c]: Done, also moved create lock up so call to cleanup will destroy it correctly.

File: mama/c_cpp/src/c/publisher.c (hunk 12 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/publisher.c]: Done

File: mama/c_cpp/src/c/publisher.c (hunk 13 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/publisher.c]: Done

File: mama/c_cpp/src/c/subscription.c (hunk 14 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/subscription.c]: This provides Windows compat for compile. Same question as mempool.c

File: mama/c_cpp/src/c/subscription.c (hunk 15 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/subscription.c]: All of these WITH_ENTITLEMENTS changes provide Window compile compat. Same question as mempool.c

File: mama/c_cpp/src/c/subscription.c (hunk 16 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/subscription.c]: All of these WITH_ENTITLEMENTS changes provide Window compile compat. Same question as mempool.c

File: mama/c_cpp/src/c/mama/publisher.h (hunk 8 above)
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: It's the argument list I'm talking about - the spaces are fine, so:

```
mamaPublisher_sendReplyToInboxHandle (mamaPublisher publisher,
                                	  mamaMsgReply  replyAddress,
                                      mamaMsgReply  replyAddress,
                                      mamaMsg       reply);
```
Rather than:
```
mamaPublisher_sendReplyToInboxHandle (mamaPublisher publisher,
                                	  mamaMsgReply replyAddress,
                                      mamaMsgReply replyAddress,
                                      mamaMsg reply);
```

File: mama/c_cpp/src/c/subscriptionimpl.h (hunk 17 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/subscriptionimpl.h]: This is for my test bridge (reflect) so it can provide a mode where an initial is guaranteed to be the first msg received.

File: common/c_cpp/src/c/mempool.c (hunk 4 above)
Comment (fquinner) [inline common/c_cpp/src/c/mempool.c]: Ideally yes, but if that's painful to retro-fit, I might be able to do some merge magic to spoof combining the two efforts instead when it comes to getting these into next.

File: mama/c_cpp/src/c/mama/publisher.h (hunk 8 above)
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: Same as above this is for parameter alignment. I know if can be painful because a lot of these files contain mixed-standards but the double-column approach is the one we try and use for any new code where all arguments won't fit on a single line (or optionally even if they would).

File: mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h (hunk 18 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h]: I changed the tabs to 4 spaces, what that the change you were looking for? The indentation pattern is the same as MamaPublisher.h, is that pattern OK?

File: mama/c_cpp/src/c/mama/publisher.h (hunk 9 above)
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: Same as above this is for parameter alignment. I know if can be painful because a lot of these files contain mixed-standards but the double-column approach is the one we try and use for any new code where all arguments won't fit on a single line (or optionally even if they would).

File: mama/c_cpp/src/c/subscription.c (hunk 14 above)
Comment (fquinner) [inline mama/c_cpp/src/c/subscription.c]: Really wish these were submitted before the publisher events stuff was started - these ones in particular will make merging the entitlements changes a nightmare. You sure there are no command line options you can use to disable the mixed declaration enforcement in the interim? This has all been compiled on windows before.

File: mama/c_cpp/src/c/subscriptionimpl.h (hunk 17 above)
Comment (fquinner) [inline mama/c_cpp/src/c/subscriptionimpl.h]: OK - does no harm was just curious.

File: mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h (hunk 18 above)
Comment (fquinner) [inline mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h]: Yup that would work thanks.

File: mama/c_cpp/src/c/mama/publisher.h (hunk 19 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/publisher.h]: Good question, we were trying to make changing the structure easier, but most people will probably just put it on the stack and assign themselves. If you want I can remove them.

File: mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp (hunk 20 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp]: The issue here is that the defaults don't run for some environments. If a tester does not have wmw installed then the unit tests will fail. I run with Solace and Tick42/RMDS for unit tests, I need to be able to specify the parms. All of the unit tests I run are scripted so adding the cmd line parms is easier.

File: mama/c_cpp/src/gunittest/c/publishertest.cpp (hunk 21 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/gunittest/c/publishertest.cpp]: Done

File: mama/dotnet/src/cs/MamaPublisher.cs (hunk 22 above)
Comment (reed-jpm-alpert) [inline mama/dotnet/src/cs/MamaPublisher.cs]: Done, tabs converted to 4 spaces

File: mama/jni/src/com/wombat/mama/MamaSubscription.java (hunk 23 above)
Comment (reed-jpm-alpert) [inline mama/jni/src/com/wombat/mama/MamaSubscription.java]: Done - will resubmit this as having a finalizer in java does not do anything since to get to finalize you have to call destroy first to clear the references, and destroy does all that finalize does.

File: mama/c_cpp/src/c/clientmanageresponder.c (hunk 1 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/clientmanageresponder.c]: Done

File: mama/c_cpp/src/c/clientmanageresponder.c (hunk 2 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/clientmanageresponder.c]: Done

File: mama/c_cpp/src/c/clientmanageresponder.c (hunk 3 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/clientmanageresponder.c]: Done

File: mama/c_cpp/src/c/mama/publisher.h (hunk 8 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/publisher.h]: Done

File: mama/c_cpp/src/c/mama/publisher.h (hunk 9 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/publisher.h]: Done

File: mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h (hunk 18 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/cpp/mama/MamaPublisherCallback.h]: Done

File: common/c_cpp/src/c/mempool.c (hunk 4 above)
Comment (reed-jpm-alpert) [inline common/c_cpp/src/c/mempool.c]: I build with scons on windows, let me check if there is a compiler arg I can use to allow this, then I can remove all of these changes.

File: mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp (hunk 20 above)
Comment (fquinner) [inline mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp]: I'm more referring to the new ones - the -m and -p ones were there already and are used to specify (for example) -m qpid -p qpidmsg so the tests can run on open source middlewares. As long as they're not mandatory for other middlewares to run though it's OK for now though.

File: mama/c_cpp/src/c/mama/publisher.h (hunk 19 above)
Comment (fquinner) [inline mama/c_cpp/src/c/mama/publisher.h]: Yeah I think letting the application decide how to allocate the callback struct is better - it's how we deal with mamaMsgCallbacks today as well. Would be a different story if the function which was receiving that pointer didn't effectively memcpy the struct anyway.

File: common/c_cpp/src/c/mempool.c (hunk 4 above)
Comment (fquinner) [inline common/c_cpp/src/c/mempool.c]: As discussed off-ticket, there is not in your version of Visual Studio. With that in mind, let's just keep these changes in your fork and I'll work out any conflicts when we merge entitlements too.

File: mama/c_cpp/src/c/subscription.c (hunk 14 above)
Comment (fquinner) [inline mama/c_cpp/src/c/subscription.c]: As discussed, this is not supported in VS2012 so going to keep this in.

**Comment by fquinner** [comment]
Thanks Reed,

It looks like something has gone funny with your branch when pulling in the latest version of next. It is all showing up as one large commit (did you squash merge?) which is making the pull request significantly bigger. If you push your current corrections as-is with all your latest changes in place, I can try and fix with an interactive merge (I think the only outstanding issue that you haven't already changed in your local copy is the _allocate / _deallocate for callbacks which we can always fix straight after it goes into next).

Cheers,
Frank

File: mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp (hunk 20 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp]: Done

**Comment by reed-jpm-alpert** [comment]
I did this in my clone of my fork.

git checkout -b OpenMAMA-next feature-publisher-events
git pull https://github.com/OpenMAMA/OpenMAMA.git next
edit conflicts
git add mama/c_cpp/src/c/publisher.c mama/c_cpp/src/gunittest/c/MainUnitTestC.cpp mama/c_cpp/src/gunittest/c/MainUnitTestC.h mama/c_cpp/src/gunittest/c/publishertest.cpp mama/jni/src/junittests/Main.java mama/jni/src/junittests/MamaPublisherTest.java
git checkout feature-publisher-events
git merge --no-ff OpenMAMA-next
git commit
git push origin feature-publisher-events

**Comment by fquinner** [comment]
Thanks Reed,

Are all of your changes pushed to your branch?

Cheers,
Frank

**Comment by reed-jpm-alpert** [comment]
Yes - the only missing is removing the allocate/deallocate for the pub msg cb struct. I can do that now - will change a number of sample files, or wait for the next branch and make it a separate change.

**Comment by fquinner** [comment]
Hi Reed - ok in that case let's leave as-is for now and I'll see what the best way is to merge this - most likely on Monday morning.

**Comment by fquinner** [comment]
Hi Reed,

This commit seems to have messed everything up: 022f0060f46e9fd14b4e16777fd51f718564d5bd

It is published as a single commit and seems to contain every change made on next since your original fork. In addition to this, the result on your branch still doesn't seem to apply cleanly to next. In an attempt to clean this up, I have done a squash commit of all your changes and applied this to a staging branch in my fork. It should contain all of your changes in isolation again, so I'll create a fresh pull request based on that and we can work from there to get this merged into next, then you can submit your allocate / deallocate changes on top of that.

**Comment by fquinner** [comment]
Hi Reed, could you review https://github.com/OpenMAMA/OpenMAMA/pull/128 please? I have put together a new clean pull request which contains all your issues which merges cleanly with next though I did encounter a few things which seemed to creep in (or were missed first time around). Could you review those please?

**Comment by fquinner** [comment]
Now merged into next (d9f874256ceaf7e9cc05d238c68628c36782cf85)
